41.md

```diff
@@ -0,0 +1,114 @@
+
+## Implementation
+
+If, at the time of first generating their Nostr public keys, one decides to use this NIP, they will have a chain of 256 keys that can be used in sequence, which means a user can lose up to 255 keys, which should be much more than enough for any user that is not actively trying to lose their keys.
+
+### Key generation
+
+This is intended to be run on safe and trusted hardware.
+
+A (BIP-32)[https://bips.xyz/32) seed is generated by any means (probably using BIP-39 words is the best idea), then the sequence of 256 hidden keys is generated with the following paths:
+
+```
+sk0' = m/44'/1237'/41'/0'
+sk1' = m/44'/1237'/41'/1'
+...
+sk2' = m/44'/1237'/41'/255'
+```
+
+- The "root" key, i.e. the last key in the sequence, is equal to the "hidden" key for the first position: `sk0` = `sk0'`.
+- Its immediate child, i.e. the second-to-last key in the sequence is calculated as follows:
+  - calculate the public key for `sk0`: `pk0 = sk0 * G`;
+  - take the bytes for the X coordinates of that: `pk0x = getX(pk0)`;
+  - calculate the public key for `sk1'`: `pk1' = sk1' * G`;
+  - take the bytes for the X coordinates of that: `pk1x' = getX(pk1')`;
+  - concatenate the bytes `pk0x` and `pk1x'` and calculate the sha256 hash of that: `hash = sha256(pk0x || pk1x')`;
+  - add that, as a number, to the hidden private key `sk1'` (which is a number): `sk1 = sk1' + hash`;
+  - the result is the second-to-last secret key `sk1`.
+- Now proceed to calculate `sk2` using the same process above, but moving everything one number up.
+- And so on and so forth until you get to `sk255`.
+- `sk255` is the key to be used first. If that gets compromised one must move to `sk254` and so on.
+
+Presumably this key generation process happens in somewhat trusted hardware by a trusted dedicated program, and from it users may copy the initial key and paste it in the normal (still trusted) client they use for day-to-day operations.
+
+### Reader implementation
+
+Clients can query relays for the invalidation of key `A` whenever they want by using the filter `{"#p": ["A"], "kinds": [13]}`.
+
+The verification process for the validity of an invalidation (`kind: 13`) event was given in the explanation above.
+
+## Additional comments
+
+### This is not a general "rotation" scheme for keys
+
+This scheme is intended to make it less catastrophic when a key is compromised because it was input into a computer, phone or Nostr client softwre that turns out later to have been compromised. It isn't intended to allow people to rotate their keys every month as a routine practice, nor is it supposed to let users be reckless and give their private keys to any malware or trusted third parties. A compromised key is still a bad event, just not one of awful and unrepairable consequences if this NIP is followed and relied upon.
+
```

Comment:
In particular, all existing events signed by the compromised key become indistinguishable from new events generated by an impostor with that key.

Reply:
Yes -- although smart stateful clients can arrange things against that.
Or, better, the key owner can republish them with dates in the past.